drivers/Aqara/aqara-lock/src/init.lua

```diff
@@ -0,0 +1,243 @@
+  ["13.46.85"] = { event_name = METHOD.KEYPAD, event_handler = HANDLERS.UNLOCK_HANDLER },
+  ["13.56.85"] = { event_name = METHOD.NO_USE, event_handler = HANDLERS.BATTERY_HANDLER },
+  ["13.32.85"] = { event_name = METHOD.NO_USE, event_handler = HANDLERS.TAMPER_ALERT_HANDLER }
+}
+
+local function request_generate_shared_key(device)
+  device:send(cluster_base.write_manufacturer_specific_attribute(device,
+    PRI_CLU, PRI_ATTR, MFG_CODE, data_types.OctetString, "\x2B"))
+end
+
+local function lock_state_handler(driver, device, value, zb_rx)
+  local shared_key = device:get_field(SHARED_KEY)
+  local param = value.value
+  local command = string.sub(param, 0, 1)
+
+  if command == "\x3E" then
+    -- recv lock_pub_key
+    local locks_pub_key = string.sub(param, 2, string.len(param))
+    local mn_id = "Id3A"
+    local setup_id = "006"
+    local product_id = ""
+    local res, _ = security.get_aqara_secret(device.zigbee_eui, locks_pub_key, "", mn_id, setup_id,
+      product_id)
+    if res then
+      print(res)
+    end
+  elseif shared_key == nil then
+    request_generate_shared_key(device)
+  elseif command == "\x93" then
+    local opts = { cipher = "aes256-ecb", padding = false }
+    local raw_key = base64.decode(shared_key)
+    local raw_data = string.sub(param, 2, string.len(param))
+    local msg = security.decrypt_bytes(raw_data, raw_key, opts)
+    local text = string.sub(msg, 5, string.len(msg))
+    local payload = string.sub(text, 4, string.len(text))
+    local func_id = toValue(payload, 1, 1) .. "." .. toValue(payload, 2, 1) .. "." .. toValue(payload, 3, 2)
+    serial_num = toValue(msg, 3, 2)
+    seq_num = string.byte(text, 3)
+
+    if resource_id[func_id] then
+      resource_id[func_id].event_handler(driver, device, resource_id[func_id].event_name,
+        toValue(payload, 6, string.byte(payload, 5)))
+    end
+
+    if serial_num >= 0xFFFF then
+      device:send(cluster_base.write_manufacturer_specific_attribute(device,
+        PRI_CLU, PRI_ATTR, MFG_CODE, data_types.OctetString, "\x2B"))
+    end
+  end
+end
+
+local function send_msg(device, funcA, funcB, funcC, op_code, length, value)
+  local shared_key = device:get_field(SHARED_KEY)
+  if shared_key == nil then
+    request_generate_shared_key(device)
+  else
+    local payload = toHex(funcA, 1) .. toHex(funcB, 1) .. toHex(funcC, 2) .. toHex(length, 1) .. toHex(value, length)
+    seq_num = seq_num + 1
+    local text = "\x00" .. toHex(op_code, 1) .. toHex(seq_num, 1) .. payload
+    serial_num = serial_num + 1
+    local raw_data = "\x5B" .. toHex(string.len(text), 1) .. toHex(serial_num, 2) .. text
+    for i = 1, 4 - (string.len(raw_data) % 4) do
+      raw_data = raw_data .. "\x00"
+    end
+
+    local opts = { cipher = "aes256-ecb", padding = false }
+    local raw_key = base64.decode(shared_key)
+    local result = security.encrypt_bytes(raw_data, raw_key, opts)
+    local msg = "\x93" .. result
```

Comment:
This is a bit complicated and might have some potential issues that would be difficult to detect at runtime. I wonder if extracting this to a helper function would be helpful with extra error logging around if any of these operations fail (for example if the inputs can't be converted to hex)